Add namespace validation for rename operation

ajantha-bhat opened the pull request:
Fixes #7645

ajantha-bhat commented on integrations/spark-extensions-basetests/src/main/java/org/projectnessie/spark/extensions/AbstractNessieSparkSqlExtensionTest.java:
```diff
@@ -132,6 +133,27 @@ public void testRefreshAfterMergeWithIcebergTableCaching()
+    // rename to the table identifier with non-existing namespace
+    assertThatThrownBy(() -> sql("ALTER TABLE nessie.db.tbl RENAME TO nessie.non_existing.table"))
+      .isInstanceOf(RuntimeException.class)
+      .hasMessage("Cannot rename table 'tbl' to 'table': the current reference is not up to date");
```
This final error from iceberg client is `CommitFailedException`, because we are using `NessieUtil.maybeThrowSpecializedException` only for commit operations. We need to fix it at the Iceberg side.

```
org.apache.iceberg.exceptions.CommitFailedException: Cannot rename table 'tbl' to 'table': the current reference is not up to date.
	at org.apache.iceberg.nessie.NessieIcebergClient.renameTable(NessieIcebergClient.java:366)
	at org.apache.iceberg.nessie.NessieCatalog.renameTable(NessieCatalog.java:283)
	at org.apache.iceberg.CachingCatalog.renameTable(CachingCatalog.java:178)
	at org.apache.iceberg.spark.SparkCatalog.renameTable(SparkCatalog.java:389)
	at org.apache.spark.sql.execution.datasources.v2.RenameTableExec.run(RenameTableExec.scala:51)
```

Actual server side error
```
Caused by: org.projectnessie.error.NessieReferenceConflictException: Namespace 'non_existing' must exist.
	at org.projectnessie.error.ErrorCode.lambda$asException$1(ErrorCode.java:66)
```

snazy commented on integrations/spark-extensions-basetests/src/main/java/org/projectnessie/spark/extensions/AbstractNessieSparkSqlExtensionTest.java:
```diff
@@ -132,6 +132,28 @@ public void testRefreshAfterMergeWithIcebergTableCaching()
   }
 
+  @Test
+  public void testRenameTableToMissingNamespace()
```
This test is way "too far away" from the changed code. A test needs to be in `:nessie-versioned-storage-common-tests` - such a test must also verify that the identified-key for `CommitValidator` works in that case.

snazy reviewed (changes requested):
Test missing / wrong place.

ajantha-bhat commented:
>  Test missing / wrong place.

I tried adding a test in `AbstractNamespaceValidation`

```
@Test
  void RenameWithNonExistingNamespace() throws Exception {
    BranchName branch = BranchName.of("renameWithNonExistingNamespace");
    store().create(branch, Optional.empty());

    Namespace ns1 = Namespace.of("ns");
    Namespace ns2 = Namespace.of("non_existing");
    ContentKey key1 = ContentKey.of(ns1, "table");
    ContentKey key2 = ContentKey.of(ns2, "tbl");

    store()
      .commit(
        branch,
        Optional.empty(),
        fromMessage("create namespace ns"),
        singletonList(
          Put.of(ns1.toContentKey(), ns1)));

    OnRefOnly content = newOnRef("value");
    store()
      .commit(
        branch,
        Optional.empty(),
        fromMessage("create a table"),
        singletonList(Put.of(key1, content)));

    store()
      .commit(
        branch,
        Optional.empty(),
        fromMessage("rename table"),
        asList(Delete.of(key1), Put.of(key2, content)));
  }
  ```
  
  But it seems it is not going to the place where I fixed. 
  
  I am not very familiar with that part of the code now. It would take too much time for me to understand it now. 
  Hence, I went with black box level testing. Agree that we should have unit tests for this. But I am not sure whether I 
  can work on it (or fix it soon).

ajantha-bhat commented:
> But it seems it is not going to the place where I fixed.

I should fetch the content from the backend to use it for rename. Without that reusing the content object in the test class will be considered as new content as it is not having content id. Fixed now.

adutra commented on versioned/tests/src/main/java/org/projectnessie/versioned/tests/AbstractNamespaceValidation.java:
```diff
@@ -509,4 +510,42 @@ void deleteHierarchy() throws Exception {
+    BranchName branch = BranchName.of("renameWithNonExistingNamespace");
+    store().create(branch, Optional.empty());
+
+    Namespace ns1 = Namespace.of("ns");
```
I think you cold simplify the test, we don't need 2 namespaces:

```java
    Namespace ns = Namespace.of("non_existing");
    ContentKey key1 = ContentKey.of("table");
    ContentKey key2 = ContentKey.of(ns, "table");

    store()
        .commit(
            branch,
            Optional.empty(),
            fromMessage("create a table"),
            singletonList(Put.of(key1, newOnRef("value"))));

    Content table = store().getValue(branch, key1).content();
// etc.
```

ajantha-bhat replied:
Done. Also added an entry for CHANGELOG.md.

adutra reviewed (commented):
Please also add an entry to CHANGELOG.md.

adutra commented:
> such a test must also verify that the identified-key for CommitValidator works in that case.

Namespace validation happens _before_ the `CommitValidator` is invoked. The identified key is correctly created, but discarded when the validation error is thrown. I think that's fine?
